APISpec.md

```diff
@@ -0,0 +1,508 @@
+
+* **Session** - A Session represents an authenticated user session and corresponds many-to-one to a User.
+
+## Data Tables
+
+### Users:
+```
+ID - Guid (PK)
+Username - String
+Password - String (Hash)
+```
+
+### Todos:
+```
+ID - Guid (PK)
+UserID - Guid (FK - ID in Users)
+Title - String
+Status - Enumified String
+CreationDate - Datetime
+DueDate - Datetime
+```
+
+### Tokens:
+```
+ID - Guid (PK)
+UserID - Guid (FK - ID in Users)
+Token - String
+Expiration - Datetime
+```
+
+## Data Models
+
+### User
+```
+ID - Guid
+Username - String
+Password - String (Hash)
+Todos - Todo[] (List of Todos for the user)
+Sessions - Session[] (List of active sessions for the user)
+```
+
+### Todo
+```
+ID - Guid
+User - User
+Title - String
+Status - Status
+CreationDate - Datetime
+DueDate - Datetime
+```
+
+### Session
+```
+ID - Guid
+Token - String (JWT from auth0)
+User - User
+Expiration - Datetime
+```
+
+### Status
+```
+OPEN
+IN_PROGRESS
+COMPLETE
+```
+
+## Endpoints
+
+### Create New Todo
+---
+
+```
+POST /todos
+```
+
+Headers:
+```
+"Authorization": Bearer <Token>
+```
+
+Body: 
+```
+{
+    "meta": {},
+    "data": {
+        "title": "Buy groceries",
+        "status": "OPEN|IN_PROGRESS|COMPLETE"
+        "dueDate": "2022-04-01T13:02:37.000Z"
+    }
+}
+```
+
+**Response** - `201 Created`
+
+Headers: 
+```
+"Location": "/todos/{id}"
+```
+
+**Error Codes**:
+
+`400 Bad Request` - Request payload invalid OR dueDate provided not a valid date
+
+`401 Unauthorized` - Request does not have valid authentication to create todo
+
+### Fetch all Todos
+---
+
+**Request**:
+``` 
+GET /todos
+```
+Headers:
+```
+"Authorization": Bearer <Token>
+```
+
+Body: None
+
+**Response** - `200 OK`
+
+Body:
+```
+{
+    "meta": {
+        "todoCount": 2
+    }, 
+    "data": [
+        {
+            "id": "8a480c40-e4be-4b89-91c9-844de091a214",
+            "title": "A todo",
+            "status": "OPEN|IN_PROGRESS|COMPLETE",
+            "creationDate": "2022-03-18T13:03:03.000Z",
+            "dueDate": "2022-04-01T00:00:00.000Z"
+        },
+        {
+            ...
+        }
+    ]
+}
+```
+
+**Error Codes**:
+
+`404 Bad Request` - Request payload invalid OR status provided not a valid enum value OR invalid dates
+
+`401 Unauthorized` - Request does not have valid authentication to fetch todos
+
+### Update a Todo
+---
+
+**Request**:
+```
+PUT /todos/{id}
+```
+
+Headers:
+```
+"Authorization": Bearer <Token>
+```
+
+Body:
+```
+{
+    "meta": {},
+    "data": {
+        "title": "My updated todo",
+        "status": "OPEN|IN_PROGRESS|COMPLETE",
+        "dueDate": "2022-04-02T00:00:00.000Z"
+    }
+}
+```
+
+**Response** - `204 No Content`
+
+**Error Codes**:
+
+`404 Not Found` - Todo with given ID does not exist in the API
+
+`400 Bad Request` - Request payload invalid OR status provided not a valid enum value
+
+`401 Unauthorized` - Request does not have valid authentication to update the todo
```

Comment:
What if a user tries to update or delete or fetch a TODO that they do not have access to?

Reply:
To clarify wouldn't a `401` be "unauthenticated" and `403` if the user does not have permission to modify the resource?

Reply:
Yes, I believe a `401` would be if the token in the `Authorization` header is missing or invalid and a `403` would be if the token is valid but the user does not have permission to modify the resource

Reply:
I added `403` where I thought was appropriate if the user does not have permission to modify the resource